deflake endpointmanager tests

bimmlerd opened the pull request:
This series of patches deflakes the endpoint manager tests. Four different flakes were observed when running just the `endpointmanager` pkg test in a loop, after this PR I achieved `31m55s: 46344 runs so far, 0 failures` which implies a high likelihood that any potentially remaining flakes are highly unlikely to occur.

There was a trivial data race in the pressure metrics _test_, fixed by using an atomic value.

This PR refactors the endpointmanager EP identifier allocation logic away from using a single, global identity pool, as that inherently causes flakes in tests sharing that pool. Specifically, there was a call to `mgr.expose` which did not check the error - the error was that in the global pool, one of the identities `[1, 3, 5, 7]` had already been allocated, and hence the exposing failed. Fixed by creating a new allocation pool for each new manager.

Further tests also didn't check `expose` errors, but these were inconsequential once the global ID pool was removed. Fixed the checks anyway, for good measure.

In addition, `TestLookup` would fail iff the EP allocated would by chance get ID `1234`.

The commits attempt to be somewhat readable in sequence, but the whole change isn't massive either.

Fixes: #26630
Fixes: #28878
Fixes: #27837

bimmlerd committed:
endpointmanager: fix data race in test

Go's race detector was unhappy with this test due to unserialised
concurrent access to the last value of the fake gauge. Use an atomic
float value instead, to ensure no weirdness can occur, and placate the
race detector.

Race detector warning (mildly edited):

WARNING: DATA RACE
Read at 0x00c000930488 by goroutine 205:
  github.com/cilium/cilium/pkg/endpointmanager.TestPolicyMapPressure.TestPolicyMapPressure.func1.func2()
      cilium/pkg/endpointmanager/policymap_pressure_test.go:27 +0x69
  github.com/stretchr/testify/assert.Eventually.func1()
      cilium/vendor/github.com/stretchr/testify/assert/assertions.go:1902 +0x33

Previous write at 0x00c000930488 by goroutine 203:
  github.com/cilium/cilium/pkg/endpointmanager.(*fakeGague).Set()
      cilium/pkg/endpointmanager/policymap_pressure_test.go:45 +0x30
  github.com/cilium/cilium/pkg/endpointmanager.(*policyMapPressure).update()
      cilium/pkg/endpointmanager/policymap_pressure.go:82 +0x32a
  github.com/cilium/cilium/pkg/endpointmanager.newPolicyMapPressure.func1()
      cilium/pkg/endpointmanager/policymap_pressure.go:57 +0x2e
  github.com/cilium/cilium/pkg/trigger.(*Trigger).waiter()
      cilium/pkg/trigger/trigger.go:201 +0x771
  github.com/cilium/cilium/pkg/trigger.NewTrigger.gowrap1()
      cilium/pkg/trigger/trigger.go:122 +0x33

Goroutine 205 (running) created at:
  github.com/stretchr/testify/assert.Eventually()
      cilium/vendor/github.com/stretchr/testify/assert/assertions.go:1902 +0x3d5
  github.com/stretchr/testify/assert.(*Assertions).Eventually()
      cilium/vendor/github.com/stretchr/testify/assert/assertion_forward.go:319 +0xc7
  github.com/cilium/cilium/pkg/endpointmanager.TestPolicyMapPressure.func1()
      cilium/pkg/endpointmanager/policymap_pressure_test.go:26 +0x2a8
  github.com/cilium/cilium/pkg/endpointmanager.TestPolicyMapPressure()
      cilium/pkg/endpointmanager/policymap_pressure_test.go:30 +0x205

Goroutine 203 (running) created at:
  github.com/cilium/cilium/pkg/trigger.NewTrigger()
      cilium/pkg/trigger/trigger.go:122 +0x36d
  github.com/cilium/cilium/pkg/endpointmanager.newPolicyMapPressure()
      cilium/pkg/endpointmanager/policymap_pressure.go:52 +0x287
  github.com/cilium/cilium/pkg/endpointmanager.TestPolicyMapPressure()
      cilium/pkg/endpointmanager/policymap_pressure_test.go:18 +0x84

Fixes: 28ce005918 (endpointmanager: fix bpf policy pressure getting stuck.)

Signed-off-by: David Bimmler <david.bimmler@isovalent.com>

bimmlerd committed:
idpool: return pointer to pool

IDPool contains a mutex, passing copies around is a potential footgun. I
don't think we ever used it incorrectly, but I don't see a reason for
all the copying either.

Signed-off-by: David Bimmler <david.bimmler@isovalent.com>

bimmlerd committed:
endpointmanager: idallocator: remove checkmate

Transform the test from using checkmate to standard Go tests, as it was
not using any of the features anyway.

Signed-off-by: David Bimmler <david.bimmler@isovalent.com>

bimmlerd committed:
endpointmanager: make idallocator a struct

The endpointmanagers idallocator package was using a package global pool
for its identifier allocation. That's fine for running the agent, but
causes flakes in testing when multiple tests access the same pool. It's
also not idiomatic Go.

This patch makes the local endpoint identifier allocator a struct, and
the next patch will move it into the endpointmanager package itself, as
there is no other consumer.

While at it, also ensure that the RemoveAll method is only called from a
testing context, by taking a testing.TB as an argument. We cannot simply
move the method into the _test.go files, as tests from other packages
use it.

Signed-off-by: David Bimmler <david.bimmler@isovalent.com>

bimmlerd committed:
endpointmanager: move EP identifier alloc into pkg

The endpoint manager assumed it was the only consumer of the idallocator
pkg anyway. Having a pkg that only has one consumer is pointless, hence
move it, including tests. This also allows unexporting everything, and
reducing API surface.

Signed-off-by: David Bimmler <david.bimmler@isovalent.com>

bimmlerd committed:
endpointmanager: deflake TestLookup

TestLookup could fail with

--- FAIL: Test (0.12s)
    --- FAIL: Test/EndpointManagerSuite (0.12s)
        --- FAIL: Test/EndpointManagerSuite/TestLookup (0.00s)
            manager_test.go:438:
                ... value *endpoint.Endpoint = &endpoint.Endpoint{[...]")
                ... Test Name: endpoint does not exist

In about 0.02% of the runs. Specifically, it would fail iff the endpoint
created happened to randomly get ID 1234, out of the pool of 4096 IDs.

Fix this by not creating an endpoint unconditionally in the test, so
that there is no chance of creating one with the "wrong" id.

Signed-off-by: David Bimmler <david.bimmler@isovalent.com>

bimmlerd committed:
endpointmanager: check expose errors in test

For some reason, lots of calls to expose didn't check the error
returned. Specifically in the test context, this is not great, as it
makes debugging test failures more difficult than necessary.

Signed-off-by: David Bimmler <david.bimmler@isovalent.com>

giorio94 commented on pkg/endpointmanager/idallocator/allocator_test.go:
```diff
@@ -6,100 +6,93 @@ package idallocator
 		// check if same ID is returned more than once
 		_, ok := idsReturned[id]
-		c.Assert(ok, Equals, false)
+		assert.False(t, ok)
```
Nit. `assert.NotContains` is probably a bit more idiomatic.

giorio94 commented on pkg/endpointmanager/idallocator/allocator_test.go:
```diff
@@ -6,100 +6,93 @@ package idallocator
 		// check if same ID is returned more than once
 		_, ok := idsReturned[id]
-		c.Assert(ok, Equals, false)
+		assert.False(t, ok)
```
Same here.

giorio94 reviewed (approved):
Thanks!

tklauser commented on pkg/endpointmanager/allocator.go:
```diff
@@ -14,21 +15,27 @@ const (
+// reallocatePool starts over with a new pool. This function is only used for
 // tests and its implementation is not optimized for production.
-func ReallocatePool() {
+func (a *epIDAllocator) reallocatePool(t testing.TB) {
```
`t` looks unused in the method. Importing `testing` in non `*_test.go` files is kind of unfortunate, so maybe we could avoid that?

bimmlerd replied:
The idea was to ensure that we don't use this method outside of the test context; but it is used (via `RemoveAll`) from other package's tests, so putting it in `_test` wasn't an option. Is there some other way we can mark this as test-only code?

tklauser replied:
Ah right. I didn't consider that this is transitively also used in other package's tests. In that case I don't think there's any other way than exposing the method in non-test code :disappointed: I know people seldomly read godoc comments and we want to prevent use of the method outside of test code, but having `testing` imports in non-test code is a bit of a code smell as well. At least having a comment on both `reallocatePool` and `RemoveAll` about the purpose of the `t` parameter would be nice.

tklauser replied:
Another option would be to `panic` on [`testing.Testing()`](https://pkg.go.dev/testing#Testing) being `true` in `RemoveAll`. That way we'd not have to pass down the `*testing.TB` (which someone could potentially still provide in production code :ninja:).

bimmlerd replied:
I've figured out a solution which allows us to just remove `RemoveAll` from non-package-local testing code, and hence I moved it to _test.go :)

bimmlerd committed:
endpointmanager: test: fix fakeGauge spelling

Signed-off-by: David Bimmler <david.bimmler@isovalent.com>

bimmlerd committed:
endpointmanager: remove RemoveAll from interface

Instead of having to import the testing pkg in non-testing code, let's
remove the blocker on just having this method in the _test.go files.

The endpoint manager in the daemon_test is re-initialized for every
test, and since we don't have package global state any more we can just
remove the call to RemoveAll, which solves all the problem nicely.

Signed-off-by: David Bimmler <david.bimmler@isovalent.com>

dylandreimerink reviewed (approved):
Datapath owned files look good to me!
